src/Essentials/src/FileSystem/FileSystem.ios.cs

```diff
@@ -158,68 +160,134 @@ protected override void Dispose(bool disposing)
+				: null;
+			Type = GetFileType(ContentType);
+			FileName = GetFileName(NameWithoutExtension, Extension);
 		}
 
 		internal override Task<Stream> PlatformOpenReadAsync()
-		{
-			Stream fileStream = File.OpenRead(FullPath);
+			=> Task.FromResult<Stream>(File.OpenRead(document.FileUrl.Path!));
 
-			return Task.FromResult(fileStream);
+		protected internal override void PlatformDispose()
+		{
+			document?.Dispose();
+			document = null;
+			assetUrl?.Dispose();
+			assetUrl = null;
+			base.PlatformDispose();
 		}
 	}
 
-	class UIImageFileResult : FileResult
+	class UIImageFileResult : MediaFileResult
 	{
-		readonly UIImage uiImage;
-		NSData data;
+		UIImage img;
+		NSDictionary metadata;
+		NSMutableData imgWithMetadata;
 
-		internal UIImageFileResult(UIImage image)
-			: base()
+		internal UIImageFileResult(UIImage img, NSDictionary metadata, string name)
 		{
-			uiImage = image;
-
-			FullPath = Guid.NewGuid().ToString() + FileExtensions.Png;
-			FileName = FullPath;
+			this.img = img;
+			this.metadata = metadata;
+			NameWithoutExtension = name;
+#pragma warning disable CA1422
+			ContentType = GetMIMEType(OldUTType.JPEG);
+			Extension = GetExtension(OldUTType.JPEG);
+#pragma warning restore CA1422
+			Type = GetFileType(ContentType);
+			FileName = GetFileName(NameWithoutExtension, Extension);
 		}
 
 		internal override Task<Stream> PlatformOpenReadAsync()
 		{
-			data ??= uiImage.AsPNG();
+			imgWithMetadata ??= GetImageWithMeta();
+			return Task.FromResult(imgWithMetadata?.AsStream());
+		}
 
-			return Task.FromResult(data.AsStream());
+		public NSMutableData GetImageWithMeta()
+		{
+			if (img == null || metadata == null)
+				return null;
+
+			using var source = CGImageSource.FromData(img.AsJPEG());
+			var destData = new NSMutableData();
+			using var destination = CGImageDestination.Create(destData, source.TypeIdentifier, 1, null);
+			destination.AddImage(source, 0, metadata);
+			destination.Close();
+			DisposeSources();
+			return destData;
+		}
+
+		protected internal override void PlatformDispose()
+		{
+			imgWithMetadata?.Dispose();
+			imgWithMetadata = null;
+			DisposeSources();
+			base.PlatformDispose();
+		}
+
+		void DisposeSources()
+		{
+			img?.Dispose();
+			img = null;
+			metadata?.Dispose();
+			metadata = null;
 		}
 	}
 
-	class PHAssetFileResult : FileResult
+	class PHPickerFileResult : MediaFileResult
 	{
-		readonly PHAsset phAsset;
+		readonly string identifier;
+		NSItemProvider provider;
 
-		internal PHAssetFileResult(NSUrl url, PHAsset asset, string originalFilename)
-			: base()
+		internal PHPickerFileResult(NSItemProvider provider)
 		{
-			phAsset = asset;
+			this.provider = provider;
+			NameWithoutExtension = provider?.SuggestedName;
+
+			identifier = GetIdentifier(provider?.RegisteredTypeIdentifiers);
 
-			FullPath = url?.AbsoluteString;
-			FileName = originalFilename;
+			if (string.IsNullOrWhiteSpace(identifier))
+				return;
```

Comment:
this will leave the object in un "unknown" state, is that what we want over throwing an exception? Will that result in an error later?